src/uwtools/tests/fixtures/gfs.yaml

```diff
@@ -11,7 +11,7 @@ fcst:
   y_grid_points: 100
   length: 12
   freq: 3
-  output_hours: '{% for h in range(0, length, freq) %}{{ "%d " % h }}{% endfor %}'
+  output_hours: '{% for h in range(0, fcst.length, fcst.freq) %}{{ "%d " % h }}{% endfor %}'
```

Comment:
This one feels like trouble...

It seems to have been the case that, during dereferencing, the "context" object used to obtain values to render Jinja2 syntax was constructed in such a way that lookups could be done either (a) By a path-from-the-top reference to a value, like the new version here, `fcst.length`; or (b) By a values-surrounding-me reference, like the old version here, `length`. The (new version) values object used by this test looks like
```
model: gfs
target: ufs-weather-model
fcst:
  horizontal_resolution: c768
  vertical_resolution: 64
  x_grid_points: 100
  y_grid_points: 100
  length: 12
  freq: 3
  output_hours: '{% for h in range(0, fcst.length, fcst.freq) %}{{ "%d " % h }}{% endfor %}'
executable: '{{ UFSEXEC }}'
undefined_env: '{{ NOPE }}'
filetype: gfs
datapath: '{{ [experiment_dir, current_cycle] | path_join }}'
filename_core: 'fv_core.res.nc'
testupdate: '{{ updatethis }}'
updatethis: 'testpassed'
grid_stats:
  total_ens_points: '{{ (grid_stats.num_ens_members * grid_stats.total_points ) // 1 }}'
  num_ens_members: 30
  points_per_level: '{{ grid_stats.total_ens_points // grid_stats.num_ens_members // fcst.vertical_resolution }}'
  ref_fcst: '{{ fcst.vertical_resolution }}'
  total_points: '{{ fcst.x_grid_points * fcst.y_grid_points * fcst.vertical_resolution }}'
```
I think this seemed to work due to the old code
```
values = {**os.environ, **ref_dict, **full_dict}
```
which would define a new values context to use for Jinja2 rendering. In the case being discussed here, `full_dict` would be the object shown above, and `ref_dict` would be the subtree
```
horizontal_resolution: c768
vertical_resolution: 64
x_grid_points: 100
y_grid_points: 100
length: 12
freq: 3
output_hours: '{% for h in range(0, fcst.length, fcst.freq) %}{{ "%d " % h }}{% endfor %}'
```
Given that, `values` would end up being: The shell environment, updated by `ref_dict`, updated by `full_dict`. Each update clobbers existing keys where they overlap. In the case of `os.environ`, that's fine. But `full_dict` clobbering keys placed at the top level by `ref_dict` is not. Take a simple example:
```
foo:
    bar: 88
    baz: "{{ bar }}"
bar: 99
```
It's ambiguous what `{{ bar }}` refers to. It will end up being `99`, because `99` is the value in what would be `full_dict`, which will clobber the `88` placed at the top level by what would be `ref_dict`.

While this local reference behavior seems convenient, it also seems to me error-prone and hard to reason about. Isn't it perhaps better to require explicit, unambiguous references in Jinja2 expressions?

Reply:
This is OBE. I have restored the global/local functionality in the updated version, and added an explicit unit test for it. I still am unsure whether this behavior is a good idea.